x/axelarnet/types/evm_wasm_translator.go

```diff
@@ -0,0 +1,87 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+
+	"github.com/axelarnetwork/utils/funcs"
+)
+
+var (
+	stringType      = funcs.Must(abi.NewType("string", "string", nil))
+	stringArrayType = funcs.Must(abi.NewType("string[]", "string[]", nil))
+	bytesType       = funcs.Must(abi.NewType("bytes", "bytes", nil))
+
+	// abi encoded bytes, with the following format:
+	// wasm method name, argument name list, argument type list, argument value list
+	payloadArguments = abi.Arguments{{Type: stringType}, {Type: stringArrayType}, {Type: stringArrayType}, {Type: bytesType}}
+)
+
+type contract struct {
+	// Contract is the address of the wasm contract
+	Contract string `json:"contract"`
+	// Msg is a json struct {"methodName": {"arg1": "val1", "arg2": "val2"}}
+	Msg map[string]interface{} `json:"msg"`
+}
+
+// wasm is the json that gets passed to the IBC memo field
+type wasm struct {
+	Wasm contract `json:"wasm"`
+}
+
+// ConstructWasmMessage creates a json serialized wasm message from Axelar defined abi encoded payload
+func ConstructWasmMessage(contractAddr string, payload []byte) ([]byte, error) {
+	args, err := payloadArguments.Unpack(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	methodName := args[0].(string)
+	argNames := args[1].([]string)
+	argTypes := args[2].([]string)
+
+	abiArguments, err := buildArguments(argTypes)
+	if err != nil {
+		return nil, err
+	}
+
+	// unpack to actual argument values
+	argValues, err := abiArguments.Unpack(args[3].([]byte))
+	if err != nil {
+		return nil, err
+	}
+
+	// convert to execute msg payload
+	executeMsg := make(map[string]interface{})
+	for i := 0; i < len(argNames); i++ {
+		executeMsg[argNames[i]] = argValues[i]
+	}
+
+	msg := wasm{
+		Wasm: contract{
+			Contract: contractAddr,
+			Msg: map[string]interface{}{
+				methodName: executeMsg,
```

Comment:
can we gmp call metadata (src chain/sender etc.) somewhere?

Reply:
can't pass to contract message itself, we could somehow add this info as packet sender